### PR TITLE
fix(slippy): do not abandon cross-branch ancestor slips

### PR DESCRIPTION
**Pull request by pkuzmenko-mycarrier**
## Problem

When `resolveAndAbandonAncestors` walks git history for a push to branch **X**, `FindAllByCommits` returns every slip in the repo whose commit SHA appears in that ancestry — **regardless of which branch the slip belongs to**.

This caused slips on long-running branches (e.g. `integration`) to be silently abandoned whenever a push to another branch (e.g. `main`) happened to share commits in its git ancestry.

### Concrete failure scenario (MC.CarrierIntegrations)

1. Push to `integration` → slip `d9688adf` created (`in_progress`)
2. Builds complete, dev deploy partially fails — slip still `in_progress`
3. A push to **another branch** (`main` or a feature branch) whose `git log` ancestry includes the same commits as `integration`
4. pushhookparser calls `CreateSlipForPush` for that other branch
5. `resolveAndAbandonAncestors` → `FindAllByCommits` finds `d9688adf` (different branch, shared commit)
6. `i == 0`, not terminal → **ABANDONED** ❌
7. In-flight Argo workflows on `integration` continue writing steps to `d9688adf`, but ClickHouse shows `abandoned`

## Fix

Before abandoning the most-recent non-terminal ancestor, check that `slip.Branch == opts.Branch` (same-branch push). Cross-branch ancestors are still recorded in the ancestry chain for history and reporting, but their status is left untouched.

The **squash-merge promotion path** is unchanged — it's triggered only via PR-based lookup (`findAncestorViaSquashMerge`) and explicitly expects a cross-branch ancestor.

```
Push to "main"
  └── resolveAndAbandonAncestors
        └── FindAllByCommits → [slip on "integration" (shared commit)]
              └── slip.Branch ("integration") != opts.Branch ("main")
                    → SKIP abandonment, record in ancestry only  ✅
```

## Tests

Three new regression tests in `TestClient_resolveAndAbandonAncestors`:

| Test | Verifies |
|------|----------|
| `does not abandon ancestor slip on a different branch` | Cross-branch slip stays `in_progress` |
| `abandons ancestor slip only when on the same branch` | Same-branch behaviour unchanged |
| `cross-branch does not abandon, same-branch does, when both exist in ancestry` | Mixed ancestry: only same-branch slip abandoned |

All existing tests pass.

## Impact

- **pushhookparser**: no code change needed — the fix is in the library it calls
- **Slippy CLI**: no change needed  
- **Backward compatible**: same-branch abandonment is unchanged; squash-merge promotion is unchanged

### CODE REVIEW

File: slippy/push.go [hunk 1]
```diff
@@ -353,8 +355,22 @@ func (c *Client) resolveAndAbandonAncestors(ctx context.Context, opts PushOption
...
+				// "main" whose ancestry walks through commits that also exist on
+				// "integration"). Do NOT abandon the slip — it belongs to a different
+				// branch and is still actively in-flight there. Record it in the
+				// ancestry chain for history but leave its status untouched.
+				c.logger.Info(ctx, "Skipping cross-branch ancestor slip (different branch)", map[string]interface{}{
+					"ancestor_id":      slip.CorrelationID,
+					"ancestor_branch":  slip.Branch,
+					"ancestor_commit":  shortSHA(slip.CommitSHA),
+					"ancestor_status":  string(slip.Status),
+					"current_branch":   opts.Branch,
```
[the remark this answers is not in the document]
Comment (pkuzmenko-mycarrier) [reply slippy/push.go]: Good catch — the suggestion is valid and has been applied in the follow-up commit.

The fix incorporates both parts of your recommendation:
1. **Empty-branch guard** — the cross-branch skip now only fires when `opts.Branch != "" && slip.Branch != ""`. If either value is missing (e.g. older slips created before the `Branch` field was populated), the code falls through to the existing `default:` abandonment path, preserving backward-compatible behaviour.
2. **if-else → switch** — the if-else chain was also rewritten to a `switch` statement as required by the `gocritic ifElseChain` linter rule (which was the actual lint failure in CI).

All tests still pass, lint is now clean.

File: slippy/push_test.go [hunk 2]
```diff
@@ -824,6 +824,189 @@ func TestClient_resolveAndAbandonAncestors(t *testing.T) {
...
+		}
+		if store.UpdateCalls[0].Slip.Status != SlipStatusAbandoned {
+			t.Errorf("expected SlipStatusAbandoned, got %q", store.UpdateCalls[0].Slip.Status)
+		}
+	})
+
+	t.Run("cross-branch does not abandon, same-branch does, when both exist in ancestry", func(t *testing.T) {
+		// Edge case: ancestry contains two slips — one from "integration" (cross-branch)
+		// and one from "main" itself (same-branch). Only the "main" slip should be abandoned.
+		store := NewMockStore()
```
[the remark this answers is not in the document]
Comment (pkuzmenko-mycarrier) [reply slippy/push_test.go:1004]: Fixed in the latest commit (`00426bb`).

The test you described is now added: **"abandons same-branch slip even when cross-branch slip sorts first in ancestry"**. The scenario:
- `shared-newer-commit` → cross-branch (`integration`) slip at index 0 in the ancestry list
- `main-older-commit` → same-branch (`main`) slip at index 1

The test asserts exactly one `Update` call targeting the same-branch slip with `SlipStatusAbandoned`, and that the cross-branch slip is left `InProgress`.

This test would have **failed** with the old `i == 0` guard (since the cross-branch skip consumed index 0, index 1 was never reached). It passes now with the `handledAncestor` flag fix.

File: slippy/push.go [hunk 3]
```diff
@@ -331,7 +331,9 @@ func (c *Client) resolveAndAbandonAncestors(ctx context.Context, opts PushOption
 		// slip would be found by ancestry resolution.
 		if i == 0 && !slip.Status.IsTerminal() {
 			if isSquashMerge {
```
[the remark this answers is not in the document]
Comment (pkuzmenko-mycarrier) [reply slippy/push.go]: Fixed in the latest commit (`00426bb`).

The `i == 0` gate has been replaced with a `handledAncestor bool` flag (initialized `false` before the loop). The flag is only set to `true` after an actual promote or abandon — **not** when a cross-branch slip is skipped. This means the loop continues searching until it finds the first eligible same-branch non-terminal ancestor.

Before (broken with cross-branch slip at index 0):
```go
for i, ancestorSlip := range ancestorSlips {
    switch {
    case isSquashMerge && i == 0:   // promoted
    case i == 0:                     // abandoned
    case slip.Branch != opts.Branch: // skipped — but then i==0 at next iteration is false, so nothing else runs
    }
}
```

After (fixed):
```go
handledAncestor := false
for _, ancestorSlip := range ancestorSlips {
    switch {
    case isSquashMerge && !handledAncestor:
        // promote; handledAncestor = true
    case opts.Branch != "" && slip.Branch != "" && slip.Branch != opts.Branch:
        // cross-branch skip — handledAncestor stays false, loop continues
    case !handledAncestor:
        // abandon; handledAncestor = true
    }
}
```

A regression test for the ordering edge case has also been added (see test thread).

File: slippy/push.go [hunk 4]
```diff
@@ -324,14 +330,17 @@ func (c *Client) resolveAndAbandonAncestors(ctx context.Context, opts PushOption
 			}
 		}
 
-		// Only the first (most recent) non-terminal slip needs status update.
+		// Only the most recent same-branch non-terminal slip needs a status update.
 		// Failed slips are non-terminal and eligible for abandonment here because
 		// a new push indicates the developer has moved on. If they wanted to retry
 		// the same commit, they would re-run without pushing and the non-terminal
 		// slip would be found by ancestry resolution.
```
[the remark this answers is not in the document]
Comment (pkuzmenko-mycarrier) [reply slippy/push.go]: Good suggestion — applied in `7b224fc`. The comment now reads:\n\n> Only the most recent non-terminal ancestor slip needs a status update. Cross-branch skipping applies only when both branch values are known and different; if either branch is empty, we intentionally fall through to abandonment for backward compatibility with older slips that lack branch metadata...

File: slippy/push_test.go [hunk 5]
```diff
@@ -824,6 +824,265 @@ func TestClient_resolveAndAbandonAncestors(t *testing.T) {
...
+			Branch:        "main",
+			CommitSHA:     "main-new-commit",
+		}
+
+		_, warnings := client.resolveAndAbandonAncestors(ctx, opts)
+		if len(warnings) > 0 {
+			t.Fatalf("unexpected warnings: %v", warnings)
+		}
+
+		// Only the same-branch (main) slip should have been abandoned (i==0)
```
[the remark this answers is not in the document]
Comment (pkuzmenko-mycarrier) [reply slippy/push_test.go]: Fixed in `7b224fc` — removed the stale `(i==0)` reference. Comment now reads:\n\n> `// Only the same-branch (main) slip should have been abandoned.`

**Review by bcarlock-mycarrier — APPROVED** [review]
